src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/DiploidSNPGenotypeLikelihoods.java

```diff
@@ -0,0 +1,453 @@
+
+        final byte observedBase1 = p1.getBase();
+        final byte qualityScore1 = qualToUse(p1, ignoreBadBases, capBaseQualsAtMappingQual, minBaseQual);
+        final byte observedBase2 = p2.getBase();
+        final byte qualityScore2 = qualToUse(p2, ignoreBadBases, capBaseQualsAtMappingQual, minBaseQual);
+
+        if ( qualityScore1 == 0 ) {
+            if ( qualityScore2 == 0 ) // abort early if we didn't see any good bases
+                return 0;
+            else {
+                return add(observedBase2, qualityScore2, (byte)0, (byte)0);
+            }
+        } else {
+            return add(observedBase1, qualityScore1, observedBase2, qualityScore2);
+        }
+    }
+
+    /**
+     *
+     * @param obsBase1    first observed base
+     * @param qual1       base qual of first observed base
+     * @param obsBase2    second observed base
+     * @param qual2       base qual of second observed base; can be 0, indicating no second base was observed for this fragment
+     * @param nObs        the number of times this quad of values was seen.  Generally 1, but reduced reads can have nObs > 1 for synthetic reads
+     * @return 0 if the base is bad, 1 otherwise
+     */
+    private int add(byte obsBase1, byte qual1, byte obsBase2, byte qual2, int nObs) {
+        // TODO-- Right now we assume that there are at most 2 reads per fragment.  This assumption is fine
+        // TODO--   given the current state of next-gen sequencing, but may need to be fixed in the future.
+        // TODO--   However, when that happens, we'll need to be a lot smarter about the caching we do here.
+
+        // Just look up the cached result if it's available, or compute and store it
+        DiploidSNPGenotypeLikelihoods gl;
+        if ( ! inCache(obsBase1, qual1, obsBase2, qual2, FIXED_PLOIDY) ) {
+            gl = calculateCachedGenotypeLikelihoods(obsBase1, qual1, obsBase2, qual2, FIXED_PLOIDY);
+        } else {
+            gl = getCachedGenotypeLikelihoods(obsBase1, qual1, obsBase2, qual2, FIXED_PLOIDY);
+        }
+
+        // for bad bases, there are no likelihoods
+        if ( gl == null )
+            return 0;
+
+        double[] likelihoods = gl.getLikelihoods();
+
+        for ( DiploidGenotype g : DiploidGenotype.values() ) {
+            double likelihood = likelihoods[g.ordinal()];
+            log10Likelihoods[g.ordinal()] += likelihood * nObs;
+        }
+
+        return 1;
+    }
+
+    private int add(byte obsBase1, byte qual1, byte obsBase2, byte qual2) {
+        return add(obsBase1, qual1, obsBase2, qual2, 1);
+    }
+
+    // -------------------------------------------------------------------------------------
+    //
+    // Dealing with the cache routines
+    //
+    // -------------------------------------------------------------------------------------
+
+    private static ThreadLocal<DiploidSNPGenotypeLikelihoods[][][][][]> CACHE = new ThreadLocal<DiploidSNPGenotypeLikelihoods[][][][][]>() {
```

Comment:
I think all this crazy code should move to a different class. How about `DiploidSNPGenotypeLikelihoodsCache`?